### ISSUE TITLE
fastq_maxee argument not applying for fastx_filter in v2.15.0

### ISSUE THREAD

**Issue by fjossandon** [torognes/vsearch#422]
Hello,
I'm processing some sequencing reads in several steps, and one of the steps uses the Expected Error to filter bad quality reads. 

Originally, I used `--fastq_mergepairs` to merge forward/reverse reads, writing the output in Fasta format with `--fastaout`, and `--fastq_eeout` to add the expected error to the header... After that I used `--fastx_filter` on the output files with `--fastq_maxee` to filter out the reads having an expected error over the threshold, but it was then when I found that Vsearch was just passing all the sequences without applying the specified filter... Only after testing and changing the output format of the merged pairs to FastQ and using `--fastq_filter` instead, the filter was correctly applied.

I assumed that any valid parameter related to the header or sequence would apply to both fastq_filter and fastx_filter, because they are shown together:
```
Trimming and filtering
  --fastx_filter FILENAME     trim and filter sequences in FASTA/FASTQ file
  --fastq_filter FILENAME     trim and filter sequences in FASTQ file
  --reverse FILENAME          FASTQ file with other end of paired-end reads
 Parameters
  --fastq_ascii INT           FASTQ input quality score ASCII base char (33)
  --fastq_maxee REAL          discard if expected error value is higher
  --fastq_maxee_rate REAL     discard if expected error rate is higher
  --fastq_maxlen INT          discard if length of sequence is longer
  --fastq_maxns INT           discard if number of N's is higher
  --fastq_minlen INT          discard if length of sequence is shorter
  --fastq_qmax INT            maximum base quality value for FASTQ input (41)
  --fastq_qmin INT            minimum base quality value for FASTQ input (0)
  --fastq_stripleft INT       delete given number of bases from the 5' end
  --fastq_stripright INT      delete given number of bases from the 3' end
  --fastq_truncee REAL        truncate to given maximum expected error
  --fastq_trunclen INT        truncate to given length (discard if shorter)
  --fastq_trunclen_keep INT   truncate to given length (keep if shorter)
  --fastq_truncqual INT       truncate to given minimum base quality
  --maxsize INT               discard if abundance of sequence is above
  --minsize INT               discard if abundance of sequence is below
```
But it does not seems the case.

This is a test case.
test.fastq:
```
@M02414:5:000000000-J8KK3:1:1101:16653:1711 1:N:0:CATCGTGA+CGTCGCTA;ee=0.6413
TTACAGAGTTGCAAAACTCCCAAACCATTGTGAACGTTACCTAAACCGTTGCTTCGGCGGGCGGCCCGGGTCCTTCCCGGCGCCCCTCGGCCATCGCGGGCGCCCGCCGGAGGAAAACCAAACTATTGCATTGTATGGCCTCTCTGAGTCTTCTGTACTGAATAAGTCAA
+
JJJJJJJJJJJJJJJJJ<JJ>JJJJJJJ;JJJJJJJJJFJJJJGDI+HJJJJJJJ3JDJJDFJF9FJJGHG38JJJ;JFF4FJ49J:JJH*JFJJJBJJEDFFJJI-5IJ<JJJJJJJJJJJGJJJJEJJJJEJJJ0EGIJ)JJJ8J0GFFJ3JJJJFJJJJJJJGFJGG
```
test.fasta
```
>@M02414:5:000000000-J8KK3:1:1101:16653:1711 1:N:0:CATCGTGA+CGTCGCTA;ee=0.6413
TTACAGAGTTGCAAAACTCCCAAACCATTGTGAACGTTACCTAAACCGTTGCTTCGGCGGGCGGCCCGGGTCCTTCCCGGCGCCCCTCGGCCATCGCGGGCGCCCGCCGGAGGAAAACCAAACTATTGCATTGTATGGCCTCTCTGAGTCTTCTGTACTGAATAAGTCAA
```

Then execute:
```
vsearch --fastq_filter test.fastq --fastq_maxee 0.5 --fastaout test_fastq_out.fasta
vsearch v2.15.0_linux_x86_64, 15.6GB RAM, 4 cores
https://github.com/torognes/vsearch

Reading input file 100%  
0 sequences kept (of which 0 truncated), 1 sequences discarded.

---------------------------------------------------------------------------------------

vsearch --fastx_filter test.fasta --fastq_maxee 0.5 --fastaout test_fasta_out.fasta
vsearch v2.15.0_linux_x86_64, 15.6GB RAM, 4 cores
https://github.com/torognes/vsearch

Reading input file 100%  
1 sequences kept (of which 0 truncated), 0 sequences discarded.
```
Using fastx_filter silently does not apply the filter and the sequence pass. I think that it should either apply it, or raise an error for using an invalid argument if applying the filter does not make sense for some reason.

Best regards,

**Comment by colinbrislawn** [comment]
This is the most well written issues I have seen in weeks! 🏆

**Comment by torognes** [comment]
Thank you for reporting this issue! I understand the problem and agree that vsearch should have raised an error when trying to apply a sequence quality filter on a FASTA file. (It will not read the expected error from the FASTA header.) Sorry for the problems caused by this. I'll update the code to detect these issues and release a new version soon.

**Comment by fjossandon** [comment]
Hello @torognes. I see, I thought that it was just reading the expected error from the header for the filtering, at least for Fasta files.  So this means that it always calculates the expected error directly from the FastQ quality line regardless of the header, which is why it cannot apply in a Fasta that don't have the quality line.

I just made a small test, manually changing the value of the `;ee=` tag in the Fastq header, and I confirm that it don't affect the filter. It ignores my header edit although the output still have the edited value, but if I use `--eeout` the output recovers the original unedited number.

 Thanks fo the clarification!

**Comment by frederic-mahe** [comment]
Here are some tests around that issue:
```sh
# fastx_filter: fastq_maxee works with fastq files
printf "@s\nA\n+\n#\n" | \
    vsearch \
        --fastx_filter - \
        --fastq_maxee 1.0 \
        --quiet \
        --fastqout /dev/null && \
    echo "success" || \
        echo "fail"

# fastx_filter: fastq_maxee doesn't work with fasta files
printf ">s\nA\n" | \
    vsearch \
        --fastx_filter - \
        --fastq_maxee 1.0 \
        --quiet \
        --fastaout /dev/null && \
    echo "fail" || \
        echo "success"
    
# fastx_filter: fastq_maxee computes ee from quality (above threshold)
# ee=1.2619
printf "@s\nAA\n+\n##\n" | \
    vsearch \
        --fastx_filter - \
        --quiet \
        --fastq_maxee 1.0 \
        --fastqout - | \
    grep -q "." && \
    echo "fail" || \
        echo "success"

# fastx_filter: fastq_maxee computes ee from quality (below threshold)
# ee=1.2619
printf "@s\nAA\n+\n##\n" | \
    vsearch \
        --fastx_filter - \
        --quiet \
        --fastq_maxee 1.4 \
        --fastqout - | \
    grep -q "." && \
    echo "success" || \
        echo "fail"

# fastx_filter: fastq_maxee computes ee from quality (not from tag)
# real ee=0.6310, tag ee=1.2, max = 1.0 -> keep read
printf "@s;ee=1.2\nA\n+\n#\n" | \
    vsearch \
        --fastx_filter - \
        --quiet \
        --fastq_maxee 1.0 \
        --fastqout - | \
    grep -q "." && \
    echo "success" || \
        echo "fail"

# fastx_filter: fastq_maxee computes ee from quality (not from tag)
# real ee=1.2619, tag ee=0.6, max = 1.0 -> discard read
printf "@s;ee=0.6\nAA\n+\n##\n" | \
    vsearch \
        --fastx_filter - \
        --quiet \
        --fastq_maxee 1.0 \
        --fastqout - | \
    grep -q "." && \
    echo "fail" || \
        echo "success"
```

**Comment by torognes** [comment]
In commit 9b6fbb54abef173e16c094d8b9550d8a2e5807f1, there is now code to detect if `fastx_filter` is used with FASTA input and any of the following options that only work with FASTQ files: `eeout`, fastq_ascii`, `fastq_eeout`, `fastq_maxee`, `fastq_maxee_rate`, `fastq_out`, `fastq_qmax`, `fastq_qmin`, `fastq_truncee`, `fastq_truncqual`,  `fastqout_discarded`, `fastqout_discarded_rev`, `fastqout_rev`. The manual is updated too.

I plan to make a new release soon.

**Comment by torognes** [comment]
@frederic-mahe, thanks for the tests! All of them pass now. Will you include them in the vsearch-test repo?

**Comment by torognes** [comment]
Fixed in version 2.15.1 just released.

**Comment by fjossandon** [comment]
Tested and working perfectly. Thanks for the quick fix and release!
